doc/GUIDE.md

```diff
@@ -0,0 +1,822 @@
+this:
+
+```c
+// mylib.h
+
+typedef int (*callback_t)(int x, int y);
+```
+
+Jextract generates the following class in a `callback_t.java` file:
+
+```java
+// callback_t.java
+
+public class callback_t {
+    public interface Function {
+        int apply(int x, int y);
+    }
+
+    public static FunctionDescriptor descriptor() { ... }
+    public static MemorySegment allocate(callback_t.Function fi, Arena arena) { ... }
+    public static int invoke(MemorySegment funcPtr,int x, int y) { ... }
+}
+```
+
+We again have a meta-data accessor for the function descriptor (`descriptor()`). There's
+an `allocate` method that can be used to allocate a new instance of this function pointer,
+who's implementation is implemented by the `fi` functional interface instance. And finally,
+there's an `invoke` method which can be used to invoke an instance of `callback_t` that
+we received from native code.
+
+For instance, let's say we have a function that accepts an instance of the `callback_t`
+function pointer type:
+
+```c
+// mylib.c
+
+int call_me_back(callback_t callback) {
+    return callback(1, 2);
+}
+```
+
+We can call this function from Java as follows:
+
+```java
+// Main.java
+
+try (Arena arena = Arena.ofConfined()) {
+    MemorySegment cb = callback_t.allocate((a, b) -> a * b, arena);
+    int result = call_me_back(cb);
+    System.out.println(result); // prints: 2
+} // 'cb' freed here
+```
+
+Here we use the lambda `(a, b) -> a * b` as the implementation of the `callback_t` instance
+we create using the call to `allocate`. This method returns an upcall stub like the ones
+returned by the `java.lang.foreign.Linker::upcallStub` method. The `arena` argument denotes
+the lifetime of the upcall stub, meaning that the upcall stub will be freed when the arena
+is closed (after which the callback instance should no longer be called).
+
+Additionally, we can using the `callback_t::invoke` method invoke an instance of
+`callback_t` that we get back from a call to a C function. Let's say we have a couple of
+functions like this:
+
+```c
+// mylib.c
+
+int mult(int x, int y) {
+    return x * y;
+}
+
+callback_t get_callback(void) {
+    return &mult;
+}
+```
+
+The `get_callback` function returns an instance of `callback_t`, which is a function pointer
+pointing to the native `mult` function. We can call `callback_t` instance that `get_callback()`
+return in Java using the `invoke` method in the `callback_t` class that jextract generates
+for us:
+
+```java
+// Main.java
+
+MemorySegment cb = get_callback();
+int result = callback_t.invoke(cb, 1, 2);
+System.out.println(result); // prints: 2
+```
+
+Here the `callback_t` instance we want to invoke is passed as the first argument to
+`invoke`, and then the `1` and `2` represent the arguments passed when calling the
+`callback_t` instance.
+
+Jextract generates function pointer classes like the `callback_t` class for function
```

Comment:
I know what you mean here - but I wonder if readers will get it w/o some kind of example.

Reply:
I'll add some examples as well.